parsers/NG.py

```diff
@@ -1,76 +1,89 @@
+# Local library imports
+import parsers.lib.config
+
+API_URL = urllib.parse.urlparse("https://niggrid.org/GenerationProfile")
+API_URL_STRING = API_URL.geturl()
+NORMALISE = {
+    "gas": "gas",
+    "gas/steam": "gas",
+    "hydro": "hydro",
+    # Per the "Electricity generation by source" plot at
+    # https://www.iea.org/countries/nigeria, the majority of Nigeria's
+    # generation comes from natural gas, so "steam" most likely implies "gas";
+    # therefore, we do not map "steam" to "unknown" (coal/gas/oil). See issues
+    # #2570 and #3651 for more information.
+    "steam": "gas",
+}
+PATTERN = re.compile(r"\((.*)\)")
+
+
+# The data is hourly, but it takes a few minutes after the turn of each hour
+# for the server to populate it. Setting the refetch frequency to 45 min will
+# ensure that if the live data is missing for a given hour when it's first
+# fetched, it will be fetched again during the same hour. (As far as I can
+# tell, the table is always populated within 15 min of the turn of the hour).
+@parsers.lib.config.refetch_frequency(datetime.timedelta(minutes=45))
+def fetch_production(zone_key="NG",
+                     session=None,
+                     target_datetime=None,
+                     logger=logging.getLogger(__name__)) -> dict:
+    """Requests the last known production mix (in MW) of a given zone."""
+    timestamp = (arrow.get(target_datetime).to("Africa/Lagos")
+                 if target_datetime is not None
+                 else arrow.now(tz="Africa/Lagos"))
+    timestamp = timestamp.replace(minute=0, second=0, microsecond=0)
+
+    # GET the landing page (HTML) and scrape some form data from it.
+    session = session or requests.Session()
+    response = session.get(API_URL_STRING)
+    soup = bs4.BeautifulSoup(response.text, "html.parser")
+    data = {tag["name"]: tag["value"] for tag in soup.find_all("input")}
+    data["ctl00$MainContent$txtReadingDate"] = timestamp.format("DD/MM/YYYY")
+    data["ctl00$MainContent$ddlTime"] = timestamp.format("HH:mm")
+
+    # Send a POST request for the desired grid data using parameters from the
+    # landing page form. The grid data is presented as an HTML table; we ignore
+    # its header and footer rows.
+    response = session.post(API_URL_STRING, data=data)
+    rows = bs4.BeautifulSoup(response.text, "html.parser").find_all("tr")[1:-1]
+    production_mix = {technology: 0.0 for technology in NORMALISE.values()}
+    for row in rows:
+        _, source, power, _ = (tag.text for tag in row.find_all("td"))
+        try:
+            technology = NORMALISE[PATTERN.search(source).group(1).casefold()]
+        except (AttributeError, KeyError) as error:
+            logger.warning(f"Unexpected source '{source.strip()}' encountered")
+            continue
+        production_mix[technology] += float(power)
+    # If we have no production from any source, assume something went wrong.
+    if all(value == 0.0 for value in production_mix.values()):
```

Comment:
Could also word this as `if sum(production_mix.values()) == 0.0:` which is a tiny bit more computationally expensive, but might read more accurately (in a "code should say what you mean" sense).

As an aside, `0 == 0.0` in Python, so `value == 0` and `value == 0.0` are equivalent. `0` is type `int` and `0.0` is type `float`, but they compare the same.

If you would like to test that the value is not too low without being 0, you can also use the centralized validation function: `parsers.lib.validation.validate(datapoint, floor=10)` for minimum 10 MW. See https://github.com/electricitymap/electricitymap-contrib/blob/master/parsers/lib/validation.py

Reply:
To be honest, I'm not sure which one better expresses the intent. I'm somewhat partial to mine, although I may be biased! I should probably use `parsers.lib.validation.validate` instead, so I suppose it doesn't make much difference either way.

Reply:
I'm fine with any of these - probably leaning towards using the `validation` lib for consistency with other parsers :+1: